examples/nlp/dialogue_state_tracking_trade.py

```diff
@@ -0,0 +1,226 @@
+parser.add_argument("--save_step_freq", default=-1, type=int)
+parser.add_argument("--optimizer_kind", default="adam", type=str)
+parser.add_argument("--amp_opt_level", default="O0", type=str, choices=["O0", "O1", "O2"])
+parser.add_argument("--shuffle_data", action='store_true')
+parser.add_argument("--num_train_samples", default=-1, type=int)
+parser.add_argument("--num_eval_samples", default=-1, type=int)
+parser.add_argument("--grad_norm_clip", type=float, default=10, help="gradient clipping")
+parser.add_argument("--teacher_forcing", default=0.5, type=float)
+args = parser.parse_args()
+
+# List of the domains to be considered
+domains = {"attraction": 0, "restaurant": 1, "taxi": 2, "train": 3, "hotel": 4}
+
+if not os.path.exists(args.data_dir):
+    raise ValueError(f'Data not found at {args.data_dir}')
+
+work_dir = f'{args.work_dir}/DST_TRADE'
+
+data_desc = MultiWOZDataDesc(args.data_dir, domains)
+
+nf = nemo_core.NeuralModuleFactory(
+    backend=nemo_core.Backend.PyTorch,
+    local_rank=args.local_rank,
+    optimization_level=args.amp_opt_level,
+    log_dir=work_dir,
+    create_tb_writer=True,
+    files_to_copy=[__file__],
+    add_time_to_log_dir=True,
+)
+
+vocab_size = len(data_desc.vocab)
+encoder = EncoderRNN(vocab_size, args.emb_dim, args.hid_dim, args.dropout, args.n_layers)
+
+decoder = nemo_nlp.nm.trainables.TRADEGenerator(
+    data_desc.vocab,
+    encoder.embedding,
+    args.hid_dim,
+    args.dropout,
+    data_desc.slots,
+    len(data_desc.gating_dict),
+    teacher_forcing=args.teacher_forcing,
+)
+
+gate_loss_fn = nemo_nlp.nm.losses.CrossEntropyLoss3D(num_classes=len(data_desc.gating_dict))
+ptr_loss_fn = nemo_nlp.nm.losses.TRADEMaskedCrossEntropy()
+total_loss_fn = nemo_nlp.nm.losses.LossAggregatorNM(num_inputs=2)
+
+
+def create_pipeline(num_samples, batch_size, num_gpus, input_dropout, data_prefix, is_training):
+    logging.info(f"Loading {data_prefix} data...")
+    shuffle = args.shuffle_data if is_training else False
+
+    data_layer = nemo_nlp.nm.data_layers.MultiWOZDataLayer(
+        args.data_dir,
+        data_desc.domains,
+        all_domains=data_desc.all_domains,
+        vocab=data_desc.vocab,
+        slots=data_desc.slots,
+        gating_dict=data_desc.gating_dict,
+        num_samples=num_samples,
+        shuffle=shuffle,
+        num_workers=0,
+        batch_size=batch_size,
+        mode=data_prefix,
+        is_training=is_training,
+        input_dropout=input_dropout,
+    )
+
+    src_ids, src_lens, tgt_ids, tgt_lens, gate_labels, turn_domain = data_layer()
+
+    data_size = len(data_layer)
+    logging.info(f'The length of data layer is {data_size}')
+
+    if data_size < batch_size:
+        logging.warning("Batch_size is larger than the dataset size")
+        logging.warning("Reducing batch_size to dataset size")
+        batch_size = data_size
+
+    steps_per_epoch = math.ceil(data_size / (batch_size * num_gpus))
```

Comment:
I saw that formula several times, not sure whether it is totally correct - are we skipping the last batch if it doesn't have "enough samples"?

Reply:
This formula is supposed to round it up, so if last batch is smaller, then it is still counted as an step.